### PR TITLE
[SYCL][RTC] Ensure template kernel instantiation

### PR DESCRIPTION
**Pull request by jopperm**
As a step towards feature parity between the `sycl` and `sycl_jit` RTC implementations, this PR cherry-picks @cperkinsintel's workaround from #16305 for triggering explicit template instantiations.

### CODE REVIEW

File: sycl/source/detail/jit_compiler.cpp [hunk 1]
```diff
@@ -1173,17 +1173,20 @@ std::vector<uint8_t> jit_compiler::compileSYCL(
...
+  // RegisteredKernelNames may contain template specializations, so we just put
+  // them in main() which ensures they are instantiated.
+  std::ostringstream ss;
+  ss << "int main() {\n";
+  for (const std::string &KernelName : RegisteredKernelNames) {
+    ss << "  (void)" << KernelName << ";\n";
   }
+  ss << "  return 0;\n}\n" << std::endl;
+
+  std::string FinalSource = SYCLSource + ss.str();
```
Comment (sommerlukas) [inline sycl/source/detail/jit_compiler.cpp:1185]: Wouldn't it be easier to create the string stream with `SYCLSource` as initial content of the stream and then simply append through the stream?

**Review by sommerlukas — COMMENTED** [review]
One minor style nit.

File: sycl/source/detail/jit_compiler.cpp (hunk 1 above)
Comment (victor-eds) [reply sycl/source/detail/jit_compiler.cpp:1185]: How difficult would it be to use actual explicit template instantiation?
```c++
template void KernelName<...>(...);
```

Comment (jopperm) [reply sycl/source/detail/jit_compiler.cpp:1185]: Very difficult, because we don't know the signature of the free-function kernel.

**Review by sommerlukas — APPROVED** [review]

**Review by cperkinsintel — APPROVED** [review]

**Comment by jopperm** [comment]
@intel/llvm-gatekeepers Please merge. The failing E2E seems unrelated to this change, and `KernelCompiler/kernel_compiler_sycl_jit.cpp` is passing. Thanks!

**Pull request merged**